File: NRDB/IceTypeStats.m
//
//  IceTypeStats.m
//  Net Deck
//
//  Created by Gereon Steffens on 16.02.14.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "IceTypeStats.h"
#import "Deck.h"

@interface IceTypeStats()
@property NSUInteger iceCount;
@end

@implementation IceTypeStats

-(IceTypeStats*) initWithDeck:(Deck *)deck
{
    if ((self = [super init]))
    {
        self.iceCount = 0;
        
        // calculate ice type distribution
        NSMutableDictionary* ice = [NSMutableDictionary dictionary];
        
        for (CardCounter* cc in deck.cards)
        {
            if (cc.card.type != NRCardTypeIce)
            {
                continue;
            }
            self.iceCount += cc.count;
            NSString* iceType = cc.card.iceType;
            
            NSNumber* n = [ice objectForKey:iceType];
            int cnt = n==nil ? 0 : [n intValue];
            n = @(cnt + cc.count);
            [ice setObject:n forKey:iceType];
        }

        NSArray* sections = [[ice allKeys] sortedArrayUsingComparator:^(NSString* s1, NSString* s2) {
            return [s1 compare:s2];
        }];
        NSMutableArray* values = [NSMutableArray array];
        
        for (NSString*s in sections)
        {
            [values addObject:[ice objectForKey:s]];
        }
        self.tableData = [[TableData alloc] initWithSections:sections andValues:values];
    }
    return self;
}

-(CPTGraphHostingView*) hostingView
{
    return [self hostingViewForDelegate:self identifier:@"Ice Type"];
}

#pragma mark - CPTPlotDataSource methods

-(CPTLayer *)dataLabelForPlot:(CPTPlot *)plot recordIndex:(NSUInteger)index
{
    static CPTMutableTextStyle *labelText = nil;
    
    if (!labelText)
    {
        labelText = [[CPTMutableTextStyle alloc] init];
        labelText.color = [CPTColor blackColor];
    }
    
    NSString* type = [self.tableData.sections objectAtIndex:index];
    NSNumber* cards = [self.tableData.values objectAtIndex:index];
    
    NSString* str = nil;
    if ([cards intValue] > 0)
    {
        float pct = [cards intValue] * 100.0 / self.iceCount;
        str = [NSString stringWithFormat:@"%@: %d\n%.1f%%", type, [cards intValue], pct];
    }
    
    // 5 - Create and return layer with label text
    return [[CPTTextLayer alloc] initWithText:str style:labelText];
}

@end

File: NetDeck/IceTypeStats.swift
//
//  IceTypeStats.swift
//  NetDeck
//
//  Created by Gereon Steffens on 25.12.16.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

import CorePlot

class IceTypeStats: Stats {
    
    private let iceCount: Int
    
    init(deck: Deck) {
        var count = 0
        var types = [String: Int]()
        for cc in deck.cards.filter({$0.card.type == .ice}) {
            count += cc.count
            
            let type = cc.card.iceType
            types[type] = (types[type] ?? 0) + cc.count
        }
        
        self.iceCount = count
        super.init()
        
        let sections = types.keys.sorted { $0 < $1 }
        let values = sections.map { types[$0]! }
        
        assert(sections.count == values.count)
        
        self.tableData = TableData(sections: sections as NSArray, andValues: values as NSArray)
    }
    
    var hostingView: CPTGraphHostingView {
        return self.hostingView(for: self, identifier: "Ice Type")
    }
    
    func dataLabel(for plot: CPTPlot, record index: UInt) -> CPTLayer? {
        let type = self.tableData.sections[Int(index)] as! String
        let cards = self.tableData.values[Int(index)] as! Int
        
        let pct = Float(cards) * 100.0 / Float(self.iceCount)
        
        let str = String(format: "%@: %d\n%.1f%%", type, cards, pct)
        return CPTTextLayer(text: str)
    }
    
}
